src/main/java/carpet/commands/PlayerCommand.java

```diff
@@ -365,6 +366,11 @@ private static int shadow(CommandContext<CommandSourceStack> context)
             Messenger.m(context.getSource(), "r Cannot shadow fake players");
             return 0;
         }
+        MinecraftServer server = Objects.requireNonNull(player.getServer());
+        if (server.isSingleplayer() && server.isSingleplayerOwner(player.getGameProfile())) {
```

Comment:
No need to check `isSinglePlayer`, dedicated server will always return `false` for `isSinglePlayerOwner`.